community/bisect.go

```diff
@@ -0,0 +1,216 @@
+// graph g and the given score function. The effort parameter determines how
+// many attempts should be made to get an improved score for any given resolution.
+func ModularScore(g graph.Graph, score func(ReducedGraph) float64, effort int, src *rand.Rand) func(float64) float64 {
+	return func(resolution float64) float64 {
+		s := math.Inf(-1)
+		for i := 0; i < effort; i++ {
+			s = math.Max(s, score(Modularize(g, resolution, src)))
+		}
+		return s
+	}
+}
+
+// SizeMultiplex is a score function that is the reciprocal of the number of communities.
+func SizeMultiplex(g ReducedMultiplex) float64 { return 1 / float64(len(g.Structure())) }
+
+// WeightMultiplex is a score function that is the sum of community weights. The concrete
+// type of g must be pointer to a ReducedUndirectedMultiplex or a ReducedDirectedMultiplex,
+// otherwise WeightMultiplex will panic.
+func WeightMultiplex(g ReducedMultiplex) float64 {
+	var w float64
+	switch g := g.(type) {
+	case *ReducedUndirectedMultiplex:
+		for _, n := range g.nodes {
+			for _, lw := range n.weights {
+				w += lw
+			}
+		}
+	case *ReducedDirectedMultiplex:
+		for _, n := range g.nodes {
+			for _, lw := range n.weights {
+				w += lw
+			}
+		}
+	default:
+		panic(fmt.Sprintf("community: invalid graph type: %T", g))
+	}
+	return w
+}
+
+// ModularMultiplexScore returns a modularized scoring function for Profile based
+// on the graph g and the given score function. The effort parameter determines how
+// many attempts should be made to get an improved score for any given resolution.
+func ModularMultiplexScore(g Multiplex, weights []float64, all bool, score func(ReducedMultiplex) float64, effort int, src *rand.Rand) func(float64) float64 {
+	return func(resolution float64) float64 {
+		s := math.Inf(-1)
+		for i := 0; i < effort; i++ {
+			s = math.Max(s, score(ModularizeMultiplex(g, weights, []float64{resolution}, all, src)))
+		}
+		return s
+	}
+}
+
+// Profile returns an approximate profile of score values in the resolution domain [low,high)
+// at the given granularity. The score is calculated by bisecting calls to fn. If log is true,
+// log space bisection is used, otherwise bisection is linear. The function fn should be
+// monotonically decreasing with increasing input in at least 1/grain samples. Profile will
+// attempt to detect non-monotonicity during the bisection.
+//
+// Since exact modularity optimization is known to be NP-hard and Profile calls modularization
+// routines repeatedly, it is unlikely to return the exact resolution profile.
+func Profile(fn func(float64) float64, log bool, grain, low, high float64) (profile []Domain, err error) {
+	if low >= high {
+		return nil, errors.New("community: zero or negative width domain")
+	}
+
+	defer func() {
+		r := recover()
+		e, ok := r.(nonDecreasing)
+		if ok {
+			err = e
+			return
+		}
+		if r != nil {
+			panic(r)
+		}
+	}()
+	profile = bisect(fn, log, grain, low, fn(low), high, fn(high))
```

Comment:
Wouldn't it be good to evaluate `fn` 1/grain times also at low and high and take the max, resp. min?